app/src/lib/stores/helpers/branch-pruner.ts

```diff
@@ -0,0 +1,153 @@
+  private timer: number | null = null
+
+  public constructor(
+    private readonly repository: Repository,
+    private readonly gitStoreCache: GitStoreCache,
+    private readonly repositoriesStore: RepositoriesStore,
+    private readonly repositoriesStateCache: RepositoryStateCache,
+    private readonly onPruneCompleted: (repository: Repository) => Promise<void>
+  ) {}
+
+  public async start() {
+    if (this.timer !== null) {
+      fatalError(
+        `A background prune task is already active and cannot begin pruning on ${
+          this.repository.name
+        }`
+      )
+    }
+
+    await this.pruneLocalBranches()
+    this.timer = window.setInterval(
+      () => this.pruneLocalBranches(),
+      BackgroundPruneMinimumInterval
+    )
+  }
+
+  public stop() {
+    if (this.timer === null) {
+      return
+    }
+
+    clearInterval(this.timer)
+    this.timer = null
+  }
+
+  private async findBranchesMergedIntoDefaultBranch(
+    repository: Repository,
+    defaultBranch: Branch
+  ): Promise<ReadonlyArray<string> | null> {
+    const gitStore = this.gitStoreCache.get(repository)
+    return (
+      (await gitStore.performFailableOperation(() =>
+        getMergedBranches(repository, defaultBranch.name)
+      )) || null
+    )
+  }
+
+  private async pruneLocalBranches(): Promise<void> {
+    if (this.repository.gitHubRepository === null) {
+      return
+    }
+
+    // Get the last time this repo was pruned
+    const lastPruneDate = await this.repositoriesStore.getLastPruneDate(
+      this.repository
+    )
+
+    // Only prune if it's been at least 24 hours since the last time
+    const dateNow = moment()
+    const threshold = dateNow.subtract(24, 'hours')
+
+    if (lastPruneDate !== null && threshold.isBefore(lastPruneDate)) {
+      log.info(
+        `Last prune took place ${moment(lastPruneDate).from(
+          dateNow
+        )} - skipping`
+      )
+      return
+    }
+
+    // Get list of branches that have been merged
+    const { branchesState } = this.repositoriesStateCache.get(this.repository)
+    const { defaultBranch } = branchesState
+
+    if (defaultBranch === null) {
+      return
+    }
+
+    const mergedBranches = await this.findBranchesMergedIntoDefaultBranch(
+      this.repository,
+      defaultBranch
+    )
+
+    if (mergedBranches === null) {
+      log.info('No branches to prune.')
+      return
+    }
+
+    // Get all branches that exist on remote
+    const localBranches = branchesState.allBranches.filter(
+      x => x.type === BranchType.Local
+    )
+
+    // Create array of branches that can be pruned
+    const branchesReadyForPruning = new Array<Branch>()
+    for (const branch of mergedBranches) {
+      const localBranch = localBranches.find(
+        localBranch =>
+          localBranch.remote !== null && localBranch.name === branch
+      )
+
+      if (
+        localBranch !== undefined &&
+        !ReservedBranches.includes(branch.toLocaleLowerCase())
```

Comment:
:art: we could filter on this earlier and save the lookup for some branches

```ts
const candidateBranches = mergedBranches.filter(b => !ReservedBranches.includes(b))

for (const branch of candidateBranches) {
  // ...
  if (localBranch) {
    branchesReadyForPruning.push(localBranch)
  }
}
```

or

```ts
for (const branch of mergedBranches) {
  if (ReservedBranches.includes(branch.toLocaleLowerCase()) {
    continue;
  }

  // ...

  if (localBranch) {
    branchesReadyForPruning.push(localBranch)
  }
}

Reply:
Good call out, I went with the first option since that reduces the runtime complexity by quite a bit!